src/main/scala/com/github/gvolpe/scalar2019/tagless/alt/dependency.scala

```diff
@@ -0,0 +1,86 @@
+package com.github.gvolpe.scalar2019.tagless.alt
+
+import cats.Functor
+import cats.implicits._
+
+/*
+ * Generalized Reader for any `F[_]` that can eliminate the environment `R` and
+ * in effect produce a `G[_]`.
+ *
+ * It abstracts over `provide` and `run`, for `ZIO` and `Kleisli`, respectively.
+ *
+ * Examples:
+ *
+ * - `Kleisli[IO, R, A] => R => IO[A]`
+ * - `TaskR[R, A] => R => Task[A]`
+ */
+abstract class GenReader[F[_], G[_], R] {
+  def apply[A]: F[A] => R => G[A]
```

Comment:
I'd call `apply` `provide` or `runReader`.

What's going too help tremendously here is adding:

```scala
def unread[A](g: G[A]): F[A]
```

Then you can start to write laws like:

```scala
provide(unread(provide(fa)(r)))(r) === provide(fa)(r)
```

Reply:
I like this, will definitely add it today :+1: